plugins/outputs/zabbix/lld.go

```diff
@@ -0,0 +1,196 @@
+	cache map[uint64]bool
+
+	// receivedData store LLD values for each host-LDDKey combination
+	receivedData map[hostLLD]lldTags
+
+	// previousReceivedData stores the receivedData of the previous period
+	previousReceivedData map[hostLLD]lldTags
+
+	// lastClear store the time of the last clear of the LLD data
+	lastClear time.Time
+
+	// clearInterval after this number of pushes, all data is considered new.
+	// The idea behind this parameter is to resend known LLDs with low freq in case
+	// previous sent was not processed by Zabbix.
+	clearInterval config.Duration
+}
+
+// lldTags stores different combinations of tags seen for a particular host and LLD
+type lldTags []map[string]string
+
+// generateDataValues create the "data" structure needed by Zabbix to send LLD
+// Eg.: {"data": [{"{#FOO}": "1", "{#BAR}": "2"}, {"{#FOO}": "3", "{#BAR}": "4"}]}
+func (l lldTags) generateDataValues() map[string][]map[string]string {
+	values := make([]map[string]string, 0, len(l))
+
+	for _, set := range l {
+		setValues := make(map[string]string, len(set))
+		for tagKey, tagValue := range set {
+			if tagKey == "host" {
+				continue
+			}
+			setValues[fmt.Sprintf("{#%s}", strings.ToUpper(tagKey))] = tagValue
+		}
+		values = append(values, setValues)
+	}
+
+	return map[string][]map[string]string{"data": values}
+}
+
+// hostLLD used as a key to identify different LLDs in the same, or different, hosts
+type hostLLD struct {
+	Hostname string
+	LLDKey   string
+}
+
+// Push returns a slice of metrics to send to Zabbix with the LLD data using the accumulated info.
+// The name of the metric will be always "lld" (const).
+// It will have only one tag, with the host.
+// It will have an uniq field, with the LLD key as the key name and the JSON data as the value
+// Eg.: lld,host=hostA disk.device.fstype.mode.path="{\"data\":[...
+func (zl *zabbixLLD) Push() []telegraf.Metric {
+	metrics := make([]telegraf.Metric, 0, len(zl.receivedData))
+
+	// Iterate over the data collected in the last period.
+	// Compare with the data sent the last time.
+	// If different, send a new LLD.
+	for key, tags := range zl.receivedData {
+		// Skip already sent LLDs
+		if zl.compareAndDelete(key, tags) {
+			continue
+		}
+
+		dataValues := tags.generateDataValues()
+
+		dataValuesJSON, err := json.Marshal(dataValues)
+		if err != nil {
+			zl.log.Warnf("Marshaling to JSON LLD tags to Zabbix format: %v", err)
+		}
+
+		// Add the LLD m
+		m := metric.New(
+			lldName,
+			map[string]string{
+				"host": key.Hostname,
+			},
+			map[string]interface{}{
+				key.LLDKey: dataValuesJSON,
+			},
+			time.Now(),
+		)
+		metrics = append(metrics, m)
+	}
+
+	// Send empty LLDs for the LLDs that were sent the last time but not this time.
+	for key := range zl.previousReceivedData {
+		emptyDataValuesJSON, err := json.Marshal(map[string]interface{}{"data": []interface{}{}})
+		if err != nil {
+			zl.log.Warnf("Marshaling to JSON empty data Zabbix format: %v", err)
+		}
```

Comment:
Same as above. If marshalling fails, how can you continue in a safe and sensible manner?

Reply:
Fixed in d4de2d4f4